File: src/de/saschahlusiak/freebloks/lobby/LobbyDialog.java
package de.saschahlusiak.freebloks.lobby;

import java.net.InetAddress;
import java.net.NetworkInterface;
import java.net.SocketException;
import java.util.ArrayList;
import java.util.Enumeration;

import de.saschahlusiak.freebloks.R;
import de.saschahlusiak.freebloks.controller.GameMode;
import de.saschahlusiak.freebloks.controller.SpielClient;
import de.saschahlusiak.freebloks.controller.SpielClientInterface;
import de.saschahlusiak.freebloks.game.CustomGameDialog;
import de.saschahlusiak.freebloks.model.Spiel;
import de.saschahlusiak.freebloks.model.Stone;
import de.saschahlusiak.freebloks.model.Turn;
import de.saschahlusiak.freebloks.network.NET_CHAT;
import de.saschahlusiak.freebloks.network.NET_SERVER_STATUS;
import de.saschahlusiak.freebloks.network.NET_SET_STONE;
import android.app.Dialog;
import android.content.Context;
import android.content.res.Configuration;
import android.os.Bundle;
import android.os.Handler;
import android.view.KeyEvent;
import android.view.View;
import android.view.Window;
import android.view.WindowManager;
import android.view.inputmethod.EditorInfo;
import android.widget.AdapterView;
import android.widget.AdapterView.OnItemClickListener;
import android.widget.AdapterView.OnItemSelectedListener;
import android.widget.EditText;
import android.widget.FrameLayout.LayoutParams;
import android.widget.GridView;
import android.widget.ListView;
import android.widget.Spinner;
import android.widget.TextView;
import android.widget.TextView.OnEditorActionListener;

public class LobbyDialog extends Dialog implements SpielClientInterface, OnItemClickListener, OnItemSelectedListener {
	SpielClient client;
	Handler handler = new Handler();
	ListView chatList;
	ChatListAdapter adapter;
	NET_SERVER_STATUS lastStatus = null;
	GridView colorGrid;
	ColorAdapter colorAdapter;
	Spinner gameMode, fieldSize;

	public LobbyDialog(Context context, OnCancelListener cancelListener, ArrayList<ChatEntry> chatEntries) {
		super(context);
		
		setCancelable(true);
		setOnCancelListener(cancelListener);
		
		requestWindowFeature(Window.FEATURE_LEFT_ICON);
		
		setContentView(R.layout.lobby_dialog);
		
		setFeatureDrawableResource(Window.FEATURE_LEFT_ICON, R.drawable.notification_waiting);

		/* to make sure we have enough real estate. not neccessary on xlarge displays */
		if ((context.getResources().getConfiguration().screenLayout & Configuration.SCREENLAYOUT_SIZE_MASK) !=
		        Configuration.SCREENLAYOUT_SIZE_XLARGE) {
			getWindow().setLayout(LayoutParams.FILL_PARENT,	LayoutParams.WRAP_CONTENT);
		}

		colorGrid = (GridView)findViewById(R.id.color_grid);
		colorAdapter = new ColorAdapter(getContext(), null, null);
		colorGrid.setAdapter(colorAdapter);

		colorGrid.setOnItemClickListener(this);
		
		gameMode = (Spinner)findViewById(R.id.game_mode);
		gameMode.setSelection(GameMode.GAMEMODE_4_COLORS_4_PLAYERS.ordinal());
		gameMode.setEnabled(false);
		gameMode.setOnItemSelectedListener(this);

		fieldSize = (Spinner)findViewById(R.id.field_size);
		fieldSize.setSelection(4);
		fieldSize.setEnabled(false);
		fieldSize.setOnItemSelectedListener(this);

		findViewById(R.id.startButton).setOnClickListener(new View.OnClickListener() {
			@Override
			public void onClick(View v) {
				LobbyDialog.this.client.request_start();
			}
		});
		findViewById(R.id.chatButton).setOnClickListener(new View.OnClickListener() {
			@Override
			public void onClick(View v) {
				sendChat();
			}
		});
		((EditText)findViewById(R.id.chatText)).setOnEditorActionListener(new OnEditorActionListener() {
			@Override
			public boolean onEditorAction(TextView v, int actionId, KeyEvent event) {
				if (actionId == EditorInfo.IME_NULL) {
					sendChat();
					return true;
				}
				return false;
			}
		});

		adapter = new ChatListAdapter(getContext(), chatEntries, GameMode.GAMEMODE_4_COLORS_4_PLAYERS);
		chatList = (ListView)findViewById(R.id.chatList);
		chatList.setAdapter(adapter);

		if (getContext().getResources().getConfiguration().orientation == Configuration.ORIENTATION_PORTRAIT)
			getWindow().setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_ADJUST_RESIZE);
	}

	@Override
	public Bundle onSaveInstanceState() {
		Bundle b = super.onSaveInstanceState();
		b.putSerializable("lastStatus", lastStatus);
		return b;
	}

	@Override
	public void onRestoreInstanceState(Bundle savedInstanceState) {
		NET_SERVER_STATUS status = (NET_SERVER_STATUS)savedInstanceState.getSerializable("lastStatus");
		if (status != null) {
			lastStatus = status;
			updateStatus();
		}
		super.onRestoreInstanceState(savedInstanceState);
	}


	public void setSpiel(SpielClient client) {
		this.client = client;
		lastStatus = null;
		/* what do we do if supplied client is null? */
		if (client == null)
			return;

		client.addClientInterface(this);
		if (!client.spiel.isStarted()) {
			/* lobby */
			findViewById(R.id.startButton).setVisibility(View.VISIBLE);
			setTitle(R.string.lobby_waiting_for_players);
		} else {
			/* chat */
			findViewById(R.id.startButton).setVisibility(View.GONE);
			setTitle(R.string.chat);
		}

	//	TextView server = (TextView)findViewById(R.id.server);
		/* TODO: this is added again and again on screen rotate */
		if (client.getLastHost() == null) {
			try {
				Enumeration<NetworkInterface> interfaces = NetworkInterface.getNetworkInterfaces();
				int n = 0;
				if (interfaces != null)
					while (interfaces.hasMoreElements()) {
						NetworkInterface i = interfaces.nextElement();
						Enumeration<InetAddress> addresses = i.getInetAddresses();
						while (addresses.hasMoreElements()) {
							InetAddress addr = addresses.nextElement();
							if (addr.isAnyLocalAddress())
								continue;
							if (addr.isLinkLocalAddress())
								continue;
							if (addr.isLoopbackAddress())
								continue;
							if (addr.isMulticastAddress())
								continue;
							String a = addr.getHostAddress();
							if (a.contains("%"))
								a = a.substring(0, a.indexOf("%"));
							adapter.add(new ChatEntry(-1, "[" + a + "]", null));
							n++;
						}
					}
				if (n == 0) /* no address found, clients will not be able to connect */
					throw new SocketException("no address found");
			//	server.setText(s);
			//	server.setTypeface(Typeface.DEFAULT_BOLD);
			//	server.setTextColor(Color.WHITE);
			} catch (SocketException e) {
				e.printStackTrace();
			//	server.setText(R.string.no_addresses_found);
			//	server.setTypeface(Typeface.DEFAULT);
			//	server.setTextColor(Color.RED);
			}
		} else {
		//	server.setTypeface(Typeface.DEFAULT);
		//	server.setText(client.getLastHost());
		}
		adapter.setGameMode(client.spiel.m_gamemode);

		updateStatus();
		adapter.notifyDataSetChanged();
	}

	void sendChat() {
		EditText edit = (EditText)findViewById(R.id.chatText);
		String s = edit.getText().toString();
		if (s.length() < 1)
			return;

		NET_CHAT chat = new NET_CHAT(s + "\n");
		LobbyDialog.this.client.send(chat);
		edit.setText("");
	}

	@Override
	protected void onStop() {
		if (client != null)
			client.removeClientInterface(this);
		super.onStop();
	}

	@Override
	public void newCurrentPlayer(int player) {

	}

	@Override
	public void stoneWillBeSet(NET_SET_STONE s) {

	}

	@Override
	public void stoneHasBeenSet(NET_SET_STONE s) {

	}

	@Override
	public void hintReceived(NET_SET_STONE s) {

	}

	@Override
	public void gameFinished() {

	}

	@Override
	public void chatReceived(final NET_CHAT c) {
		chatList.postDelayed(new Runnable() {
			@Override
			public void run() {
				adapter.notifyDataSetChanged();
			}
		}, 100);
	}

	@Override
	public void gameStarted() {
		dismiss();
	}

	@Override
	public void stoneUndone(Turn t) {

	}

	@Override
	public void serverStatus(final NET_SERVER_STATUS status) {
		handler.postDelayed(new Runnable() {
			@Override
			public void run() {
				lastStatus = status;
				adapter.setGameMode(client.spiel.m_gamemode);
				updateStatus();
				adapter.notifyDataSetChanged();
			}
		}, 100);
	}

	void updateStatus() {
		/* better: dismiss */
		if (client == null)
			return;
		colorAdapter.setCurrentStatus(client.spiel, lastStatus);

		TextView clients = (TextView)findViewById(R.id.clients);
		if (clients != null) {
			if (lastStatus == null) {
				findViewById(R.id.clients).setVisibility(View.INVISIBLE);
			} else {
				findViewById(R.id.clients).setVisibility(View.VISIBLE);
				((TextView)findViewById(R.id.clients)).setText(getContext().getString(R.string.connected_clients, lastStatus.clients));
			}
		}
		
		if (lastStatus != null) {
			gameMode.setSelection(lastStatus.gamemode.ordinal());
			gameMode.setEnabled(!client.spiel.isStarted() && lastStatus.isVersion(3));
			
			int slider = 3;
			for (int i = 0; i < CustomGameDialog.FIELD_SIZES.length; i++)
				if (CustomGameDialog.FIELD_SIZES[i] == lastStatus.width)
					slider = i;
			fieldSize.setSelection(slider);
			fieldSize.setEnabled(!client.spiel.isStarted() && lastStatus.isVersion(3));

		} else {
			gameMode.setEnabled(false);
			fieldSize.setEnabled(false);
		}
	}

	@Override
	public void onConnected(Spiel spiel) {

	}

	@Override
	public void onDisconnected(Spiel spiel) {
		dismiss();
	}

	@Override
	public void onItemClick(AdapterView<?> adapter, View view, int position, long id) {
		if (client.spiel.isStarted())
			return;
		if (client.spiel.is_local_player((int)id)) {
			client.revoke_player((int)id);
		} else {
			client.request_player((int)id, null);
		}
	}

	@Override
	public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
		final int[] stones_default = {
			1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1
		};
		final int stones_junior[] = {
			2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 0, 0, 0, 0, 0, 0, 0, 2, 2, 0, 0
		};
		GameMode g = GameMode.from(gameMode.getSelectedItemPosition());
		if (lastStatus == null)
			return;
		int size = CustomGameDialog.FIELD_SIZES[fieldSize.getSelectedItemPosition()];
		
		client.request_game_mode(size, size, g, g == GameMode.GAMEMODE_JUNIOR ? stones_junior : stones_default);
	}

	@Override
	public void onNothingSelected(AdapterView<?> parent) {
		
	}
}

File: src/main/java/de/saschahlusiak/freebloks/lobby/LobbyDialog.kt
package de.saschahlusiak.freebloks.lobby

import android.app.AlertDialog
import android.app.Dialog
import android.content.res.Configuration
import android.text.Editable
import android.text.TextWatcher
import android.view.View
import android.view.Window
import android.view.WindowManager.LayoutParams.*
import android.view.inputmethod.EditorInfo
import android.widget.*
import android.widget.AdapterView.OnItemClickListener
import android.widget.AdapterView.OnItemSelectedListener
import android.widget.TextView.OnEditorActionListener
import androidx.lifecycle.Observer
import androidx.preference.PreferenceManager
import de.saschahlusiak.freebloks.R
import de.saschahlusiak.freebloks.client.GameClient
import de.saschahlusiak.freebloks.client.GameEventObserver
import de.saschahlusiak.freebloks.game.CustomGameDialog
import de.saschahlusiak.freebloks.game.FreebloksActivity
import de.saschahlusiak.freebloks.game.GameConfiguration
import de.saschahlusiak.freebloks.model.GameMode
import de.saschahlusiak.freebloks.model.GameMode.Companion.from
import de.saschahlusiak.freebloks.network.message.MessageServerStatus
import kotlinx.android.synthetic.main.edit_name_dialog.view.*
import kotlinx.android.synthetic.main.lobby_dialog.*

class LobbyDialog(private val activity: FreebloksActivity) : Dialog(activity), GameEventObserver, OnItemClickListener {
    private val viewModel = activity.viewModel

    private val chatAdapter = ChatListAdapter(activity, GameMode.GAMEMODE_4_COLORS_4_PLAYERS)
    private val colorAdapter = ColorAdapter(this, context, null, null)

    private var client: GameClient? = null

    private val chatObserver = Observer<List<ChatEntry>> { chatEntries ->
        chatAdapter.setData(chatEntries)
    }

    private val gameModeSelectedListener = object : OnItemSelectedListener {
        override fun onNothingSelected(parent: AdapterView<*>?) {}

        override fun onItemSelected(parent: AdapterView<*>?, view: View?, position: Int, id: Long) {
            val client = client ?: return

            val gameMode = from(game_mode.selectedItemPosition)

            val size = when (gameMode) {
                GameMode.GAMEMODE_DUO,
                GameMode.GAMEMODE_JUNIOR -> 14

                GameMode.GAMEMODE_2_COLORS_2_PLAYERS -> 15

                else -> 20
            }

            val stones = when (gameMode) {
                GameMode.GAMEMODE_JUNIOR -> GameConfiguration.JUNIOR_STONE_SET
                else -> GameConfiguration.DEFAULT_STONE_SET
            }

            client.requestGameMode(size, size, gameMode, stones)
        }
    }

    private val sizeSelectedListener = object : OnItemSelectedListener {
        override fun onNothingSelected(parent: AdapterView<*>?) {}

        override fun onItemSelected(parent: AdapterView<*>?, view: View?, position: Int, id: Long) {
            val client = client ?: return
            val gameMode = client.game.gameMode

            val size = CustomGameDialog.FIELD_SIZES[field_size.selectedItemPosition]

            val stones = when (gameMode) {
                GameMode.GAMEMODE_JUNIOR -> GameConfiguration.JUNIOR_STONE_SET
                else -> GameConfiguration.DEFAULT_STONE_SET
            }

            client.requestGameMode(size, size, gameMode, stones)
        }
    }

    init {
        setCancelable(true)

        requestWindowFeature(Window.FEATURE_LEFT_ICON)
        setContentView(R.layout.lobby_dialog)
        setFeatureDrawableResource(Window.FEATURE_LEFT_ICON, R.drawable.notification_waiting_large)

        /* to make sure we have enough real estate. not necessary on xlarge displays */
        if (activity.resources.configuration.screenLayout and Configuration.SCREENLAYOUT_SIZE_MASK !=
            Configuration.SCREENLAYOUT_SIZE_XLARGE) {
            window?.setLayout(FrameLayout.LayoutParams.MATCH_PARENT, FrameLayout.LayoutParams.WRAP_CONTENT)
        }

        if (context.resources.configuration.orientation == Configuration.ORIENTATION_PORTRAIT) {
            window?.setSoftInputMode(SOFT_INPUT_ADJUST_RESIZE or SOFT_INPUT_STATE_HIDDEN)
        } else {
            window?.setSoftInputMode(SOFT_INPUT_ADJUST_PAN or SOFT_INPUT_STATE_HIDDEN)
        }

        color_grid.apply {
            adapter = colorAdapter
            onItemClickListener = this@LobbyDialog
        }

        game_mode.apply {
            setSelection(GameMode.GAMEMODE_4_COLORS_4_PLAYERS.ordinal)
            isEnabled = false
            onItemSelectedListener = gameModeSelectedListener
        }

        field_size.apply {
            setSelection(4)
            isEnabled = false
            onItemSelectedListener = sizeSelectedListener
        }

        startButton.setOnClickListener { client?.requestGameStart() }

        chatText.apply {
            addTextChangedListener(object : TextWatcher {
                override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
                override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {}
                override fun afterTextChanged(s: Editable) {
                    chatButton.isEnabled = s.isNotEmpty()
                }
            })
            setOnEditorActionListener(OnEditorActionListener { _, actionId, _ ->
                if (actionId == EditorInfo.IME_NULL) {
                    sendChat()
                    return@OnEditorActionListener true
                }
                false
            })
        }

        chatButton.apply {
            isEnabled = false
            setOnClickListener { sendChat() }
        }

        chatList.adapter = chatAdapter
    }

    override fun onStart() {
        super.onStart()
        val client = viewModel.client
        if (client == null) {
            dismiss()
            return
        }

        this.client = client

        client.addObserver(this)

        if (client.game.isStarted) {
            /* chat */
            startButton.visibility = View.GONE
            setTitle(R.string.chat)
            setCanceledOnTouchOutside(true)
        } else {
            /* lobby */
            startButton.visibility = View.VISIBLE
            setTitle(R.string.lobby_waiting_for_players)
            setCanceledOnTouchOutside(false)
        }

        chatAdapter.gameMode = client.game.gameMode
        updateViewsFromStatus()
        chatAdapter.notifyDataSetChanged()

        viewModel.chatHistoryAsLiveData.observe(activity, chatObserver)
    }

    override fun onStop() {
        client?.removeObserver(this)
        viewModel.chatHistoryAsLiveData.removeObserver(chatObserver)
        super.onStop()
    }

    fun editPlayerName(player: Int) {
        val lastStatus = client?.lastStatus ?: return

        val dialogBuilder = AlertDialog.Builder(context)
        val inflater = this.layoutInflater
        val dialogView = inflater.inflate(R.layout.edit_name_dialog, null)
        dialogBuilder.setView(dialogView)
        val edit = dialogView.edit
        edit.setText(lastStatus.getClientName(lastStatus.getClient(player)))
        dialogBuilder.setTitle(R.string.prefs_player_name)
        dialogBuilder.setPositiveButton(android.R.string.ok) { _, _ ->
            val name = edit.text.toString().trim()
            PreferenceManager.getDefaultSharedPreferences(context)
                .edit()
                .putString("player_name", name)
                .apply()

            client?.revokePlayer(player)
            client?.requestPlayer(player, name)
        }
        dialogBuilder.setNegativeButton(android.R.string.cancel) { _, _ -> }

        val b = dialogBuilder.create()
        b.show()

        edit.selectAll()
        edit.requestFocus()
        b.window?.clearFlags(FLAG_NOT_FOCUSABLE or FLAG_ALT_FOCUSABLE_IM)
        b.window?.setSoftInputMode(SOFT_INPUT_STATE_ALWAYS_VISIBLE)
    }

    /**
     * Send what ever is currently in the chatText view to the server and reset the edit field
     */
    private fun sendChat() {
        val text = chatText.text.toString()
        if (text.isEmpty()) return
        client?.sendChat(text)
        chatText.setText("")
    }

    override fun gameStarted() {
        dismiss()
    }

    override fun serverStatus(status: MessageServerStatus) {
        chatList.post { updateViewsFromStatus() }
    }

    override fun playerJoined(client: Int, player: Int, name: String?) {
        chatList.post { updateViewsFromStatus() }
    }

    override fun playerLeft(client: Int, player: Int, name: String?) {
        chatList.post { updateViewsFromStatus() }
    }

    override fun onDisconnected(client: GameClient, error: Exception?) {
        dismiss()
    }

    private fun updateViewsFromStatus() {
        /* better: dismiss */
        val client = client ?: return
        val status = client.lastStatus

        colorAdapter.setCurrentStatus(client.game, status)

        if (status == null) {
            clients?.visibility = View.INVISIBLE
            game_mode.isEnabled = false
            field_size.isEnabled = false
        } else {
            chatAdapter.gameMode = status.gameMode

            clients?.visibility = View.VISIBLE
            clients?.text = context.resources.getQuantityString(R.plurals.connected_clients, status.clients, status.clients)

            game_mode.setSelection(status.gameMode.ordinal)
            game_mode.isEnabled = !client.game.isStarted

            var slider = 3
            for (i in CustomGameDialog.FIELD_SIZES.indices)
                if (CustomGameDialog.FIELD_SIZES[i] == status.width)
                    slider = i

            field_size.setSelection(slider)
            field_size.isEnabled = !client.game.isStarted
        }
    }

    override fun onItemClick(adapter: AdapterView<*>?, view: View?, position: Int, id: Long) {
        val client = client ?: return
        if (client.game.isStarted) return

        if (client.game.isLocalPlayer(id.toInt())) {
            client.revokePlayer(id.toInt())
        } else {
            client.requestPlayer(id.toInt(), null)
        }
    }
}